cashews/wrapper/transaction.py

```diff
@@ -0,0 +1,410 @@
+import asyncio
+import time
+from contextvars import ContextVar
+from enum import Enum
+from functools import wraps
+from typing import Any, AsyncIterator, Mapping, Optional, Tuple
+from uuid import uuid4
+
+from cashews import LockedError
+from cashews._typing import AsyncCallable_T, Middleware
+from cashews.backends.interface import Backend
+from cashews.backends.memory import Memory
+
+_transaction = ContextVar("transaction", default=None)
+_empty = object()
+_GLOBAL_LOCK_KEY = "cashews:serializable:lock"
+
+
+class TransactionMode(Enum):
+    FAST = "fast"  # simple inmemory impl, allow to have lost updates,
+    LOCKED = "locked"  # lock per key - not allow to work in parallel with the same key
+    SERIALIZABLE = "serializable"  # global lock - not allow any parallel changes
+
+
+async def transaction(
+    mode: Optional[TransactionMode] = None, timeout: Optional[float] = None
+) -> "TransactionContextDecorator":
+    return TransactionContextDecorator(mode, timeout)
+
+
+class TransactionWrapperMixin:
+    transaction_mode = TransactionMode.LOCKED
+    transaction_timeout = 10
+
+    def set_transaction_timeout(self, timeout: int):
+        self.transaction_timeout = timeout
+
+    def set_transaction_mode(self, mode: TransactionMode):
+        self.transaction_mode = mode
+
+    def _get_backend_and_config(self, key: str) -> Tuple[Backend, Tuple[Middleware, ...]]:
+        backend, config = super()._get_backend_and_config(key)
+        tx: Optional[Transaction] = _transaction.get()
+        if not tx:
+            return backend, config
+        return tx.wrap(backend), config
+
+    def transaction(
+        self, mode: Optional[TransactionMode] = None, timeout: Optional[float] = None
+    ) -> "TransactionContextDecorator":
+        mode = mode or self.transaction_mode
+        timeout = timeout or self.transaction_timeout
+        return TransactionContextDecorator(mode, timeout)
+
+
+class TransactionContextDecorator:
+    def __init__(self, mode: Optional[TransactionMode] = None, timeout: Optional[float] = None):
+        self._mode = mode
+        self._timeout = timeout
+
+    @property
+    def current_tx(self) -> Optional["Transaction"]:
+        return _transaction.get()
+
+    async def __aenter__(self) -> "Transaction":
+        if self.current_tx:
+            # warning that nested transaction not supported
+            return self.current_tx
+        return self.start()
+
+    def start(self) -> "Transaction":
+        tx = Transaction(self._mode, self._timeout)
+        _transaction.set(tx)
+        return tx
+
+    async def __aexit__(self, exc_type, exc_value, exc_tb):
+        if not self.current_tx:
+            return
+        if not exc_tb:
+            await self.current_tx.commit()
+        else:
+            await self.current_tx.rollback()
+        _transaction.set(None)
+        return False
```

Comment:
That's the way to raise an exception  - part of context manager protocol

Reply:
Seems like you can just omit return here. Initially, I asked because you have inconsistent return here - you have a simple `return` a few line above and then `return False` here.

Reply:
yep you right thanks